orca/orca.go

```diff
@@ -0,0 +1,175 @@
+ * limitations under the License.
+ */
+
+// Package orca implements Open Request Cost Aggregation, which is an open
+// standard for request cost aggregation and reporting by backends and the
+// corresponding aggregation of such reports by L7 load balancers (such as
+// Envoy) on the data plane. In a proxyless world with gRPC enabled
+// applications, aggregation of such reports will be done by the gRPC client.
+//
+// Experimental
+//
+// Notice: All APIs is this package are EXPERIMENTAL and may be changed or
+// removed in a later release.
+package orca
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"time"
+
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/grpclog"
+	"google.golang.org/grpc/metadata"
+	"google.golang.org/protobuf/proto"
+
+	v3orcapb "github.com/cncf/xds/go/xds/data/orca/v3"
+	v3orcaservicegrpc "github.com/cncf/xds/go/xds/service/orca/v3"
+)
+
+var logger = grpclog.Component("orca-backend-metrics")
+
+// EnableMetricsReportingForUnaryRPCs returns a server interceptor which enables
+// the reporting of per-RPC custom backend metrics for unary RPCs.
+//
+// Server applications interested in injecting custom backend metrics should
+// install this interceptor when creating a gRPC server, preferably as the first
+// interceptor in their chain, using grpc.ChainUnaryInterceptor().
+//
+// Subsequently, server RPC handlers can retrieve a reference to the RPC
+// specific custom metrics recorder [CallMetricRecorder] to be used, via a call
+// to GetCallMetricRecorder(), and inject custom metrics at any time during the
+// RPC lifecycle.
+//
+// The injected custom metrics will be sent as part of trailer metadata, as a
+// binary-encoded [ORCA LoadReport] protobuf message, with the metadata key
+// being set be "endpoint-load-metrics-bin".
+//
+// [ORCA LoadReport]: (https://github.com/cncf/xds/blob/main/xds/data/orca/v3/orca_load_report.proto#L15)
+func EnableMetricsReportingForUnaryRPCs() grpc.UnaryServerInterceptor {
+	return func(ctx context.Context, req interface{}, _ *grpc.UnaryServerInfo, handler grpc.UnaryHandler) (interface{}, error) {
+		r := &CallMetricRecorder{recorder: newMetricRecorder()}
+		ctxWithRecorder := setCallMetricRecorder(ctx, r)
+		resp, err := handler(ctxWithRecorder, req)
+		if err2 := r.recorder.setTrailerMetadata(ctx); err2 != nil {
+			logger.Warning(err2)
+		}
+		return resp, err
+	}
+}
+
+// wrappedStream wraps the grpc.ServerStream received by the streaming
+// interceptor. Overrides only the Context() method to return a context which
+// contains a reference to the CallMetricRecorder corresponding to this stream.
+type wrappedStream struct {
+	grpc.ServerStream
+	ctx context.Context
+}
+
+func (w *wrappedStream) Context() context.Context {
+	return w.ctx
+}
+
+// EnableMetricsReportingForStreamingRPCs is similar to
+// EnableReportingForUnaryRPCs, except that is it to be used for streaming RPCs.
+func EnableMetricsReportingForStreamingRPCs() grpc.StreamServerInterceptor {
```

Comment:
How about:

```go
func ServerMetricsReporting/*??*/() grpc.ServerOption {
	return join(grpc.ChainStreamInterceptor(...), grpc.ChainUnaryInterceptor(...))
}
```

OK so what's that `join`?  We could have an internal-only `ServerOption` that combines any arbitrary server options, or would could even export it for users.

Either way, I think something like that would be a much nicer API.

Reply:
Done.